File: service-car-rental/src/main/java/com/acme/services/carRental/java/CarRentalServiceImpl.java
package com.acme.services.carRental.java;

import com.acme.contracts.services.carRental.CarRentalCart;
import com.acme.contracts.services.carRental.CarRentalResult;
import com.acme.contracts.services.carRental.CarRentalService;
import com.acme.common.AbstractService;
import com.acme.common.ExponentialBackoffRetry;
import io.infinitic.annotations.Retry;
import org.jetbrains.annotations.NotNull;

import java.text.SimpleDateFormat;


@SuppressWarnings("unused")
@Retry(with = ExponentialBackoffRetry.class)
public class CarRentalServiceImpl extends AbstractService implements CarRentalService {
    private final SimpleDateFormat dateFormatter = new SimpleDateFormat("hh:mm:ss.SSS");

    @Override @NotNull
    public CarRentalResult book(CarRentalCart cart) {
        // fake emulation of success/failure
        log("car rental started...");

        long r = (long) (Math.random() * 5000);
        try {
            Thread.sleep(r);
        } catch (InterruptedException e) {
            throw new RuntimeException("interrupted");
        }

        if (r >= 4000) {
           log("car rental request rejected");
            return CarRentalResult.REJECTED;
        }

        // Uncomment those lines to emulate failures and retries
//        if (r >= 3800 ) {
//            log("car rental service failed!");
//            throw new RuntimeException("failing request");
//        }

        log("car rental succeeded");
        return CarRentalResult.SUCCESS;
    }

    @Override
    public void rollback(CarRentalCart cart) {
        log("car rental request rolled back");
    }

}

File: service-car-rental/src/main/kotlin/com/acme/services/carRental/kotlin/CarRentalServiceImpl.kt
package com.acme.services.carRental.kotlin

import com.acme.contracts.services.carRental.CarRentalCart
import com.acme.contracts.services.carRental.CarRentalResult
import com.acme.contracts.services.carRental.CarRentalService
import com.acme.common.AbstractService.log
import com.acme.common.ExponentialBackoffRetry
import io.infinitic.annotations.Batch
import io.infinitic.annotations.Retry
import kotlin.random.Random


@Suppress("unused")
@Retry(with = ExponentialBackoffRetry::class)
class CarRentalServiceImpl : CarRentalService {

     override fun book(cart: CarRentalCart): CarRentalResult {
        log("car rental started...")

        // fake emulation of success/failure
        val r = Random.nextLong(0, 5000)

        return when {
            r >= 4000 -> {
                log("car rental request rejected")
                CarRentalResult.REJECTED
            }

//            r >= 3800 -> {
//                log("car rental service failed!")
//                throw RuntimeException("failing request")
//            }

            else -> {
                log("car rental succeeded")
                CarRentalResult.SUCCESS
            }
        }
    }

    override fun rollback(cart: CarRentalCart) {
       log("car rental request rolled back")
    }
}
